src/server/common/cssParser.ts

```diff
@@ -0,0 +1,205 @@
+
+  let pos = 0;
+
+  function unexpected() {
+    return new Error(`Unexpected token "${tokens[pos].toSource()}" while parsing selector "${selector}"`);
+  }
+
+  function skipWhitespace() {
+    while (tokens[pos] instanceof css.WhitespaceToken)
+      pos++;
+  }
+
+  function isIdent(p = pos) {
+    return tokens[p] instanceof css.IdentToken;
+  }
+
+  function isString(p = pos) {
+    return tokens[p] instanceof css.StringToken;
+  }
+
+  function isNumber(p = pos) {
+    return tokens[p] instanceof css.NumberToken;
+  }
+
+  function isComma(p = pos) {
+    return tokens[p] instanceof css.CommaToken;
+  }
+
+  function isCloseParen(p = pos) {
+    return tokens[p] instanceof css.CloseParenToken;
+  }
+
+  function isStar(p = pos) {
+    return (tokens[p] instanceof css.DelimToken) && tokens[p].value === '*';
+  }
+
+  function isEOF(p = pos) {
+    return tokens[p] instanceof css.EOFToken;
+  }
+
+  function isClauseCombinator(p = pos) {
+    return (tokens[pos] instanceof css.DelimToken) && (['>', '+', '~'].includes(tokens[pos].value));
+  }
+
+  function isSelectorClauseEnd(p = pos) {
+    return isComma(p) || isCloseParen(p) || isEOF(p) || isClauseCombinator(p) || (tokens[p] instanceof css.WhitespaceToken);
+  }
+
+  function consumeSelectorList(): ParsedSelectorList {
+    const result = [consumeSelector()];
+    while (true) {
+      skipWhitespace();
+      if (!isComma())
+        break;
+      pos++;
+      result.push(consumeSelector());
+    }
+    return result;
+  }
+
+  function consumeSelector(): ParsedSelector | ParsedSelectorLiteral {
+    skipWhitespace();
+    const result = { clauses: [{ clause: consumeSelectorClause(), combinator: '' as ClauseCombinator }] };
+    while (true) {
+      skipWhitespace();
+      if (isClauseCombinator()) {
+        result.clauses[result.clauses.length - 1].combinator = tokens[pos++].value as ClauseCombinator;
+        skipWhitespace();
+      } else if (isSelectorClauseEnd()) {
+        break;
+      }
+      result.clauses.push({ combinator: '', clause: consumeSelectorClause() });
+    }
+    if (result.clauses.length === 1 && typeof result.clauses[0].clause === 'string')
+      return result.clauses[0].clause;
+    return result;
+  }
+
+  function consumeSelectorClause(): ParsedSelectorClause {
+    // TODO: Consider symbols like `=`, `|=`, `~=`, `*=`, `/` and convert them to strings.
+    if ((isNumber() || isString() || isStar() || isIdent()) && isSelectorClauseEnd(pos + 1))
+      return isString() ? tokens[pos++].value : tokens[pos++].toSource();
+
+    let rawCSSString = '';
+    const funcs: { name: string, args: ParsedSelectorList }[] = [];
+
+    while (!isSelectorClauseEnd()) {
+      if (isIdent() || isStar()) {
+        rawCSSString += tokens[pos++].toSource();
+      } else if (tokens[pos] instanceof css.HashToken) {
+        rawCSSString += tokens[pos++].toSource();
+      } else if ((tokens[pos] instanceof css.DelimToken) && tokens[pos].value === '.') {
+        pos++;
```

Comment:
can we have a helper function `advance()` that'll check for out-of-bounds for `pos`? I think most of the code here will work since running `instanceof` against `undefined` is fine, but it feels shady.